src/main/java/org/mqttbee/mqtt/handler/MqttChannelInitializer.java

```diff
@@ -67,6 +103,46 @@ protected void initChannel(final SocketChannel channel) {
 
         pipeline.addLast(MqttConnectHandler.NAME, new MqttConnectHandler(connect, connAckEmitter, clientData));
         pipeline.addLast(MqttDisconnectHandler.NAME, channelComponent.disconnectHandler());
+
+    }
+
+    private void initMqttOverWebSockets(@NotNull final ChannelPipeline pipeline) {
+        try {
+            final URI serverUri = new URI(
+                    getWebsocketScheme(), null,
+                    clientData.getServerHost(), clientData.getServerPort(), "/" + clientData.getServerPath(), null, null);
+            final MqttWebSocketClientProtocolHandler wsProtocolHandler = new MqttWebSocketClientProtocolHandler(serverUri);
+
+            wsProtocolHandler.onWebSocketHandshakeComplete(() -> {
+                pipeline.addLast(WS_BINARYFRAME_CREATOR, new WebSocketBinaryFrameEncoder());
+                pipeline.addLast(WS_BINARYFRAME_UNPACKER, new WebSocketBinaryFrameDecoder());
+                this.initMqttHandlers(pipeline);
+            });
+
+            pipeline.addLast(HTTP_CODEC, new HttpClientCodec());
+            pipeline.addLast(HTTP_AGGREGATOR, new HttpObjectAggregator(MAXIMUM_PACKET_SIZE_LIMIT));
+            pipeline.addLast(WEBSOCKETS_CLIENT, wsProtocolHandler);
+
+        } catch (URISyntaxException e) {
+            Exceptions.propagate(e);
+        }
+    }
+
+    private void initSsl(@NotNull final Channel channel) {
+        try {
+            //create a new SslHandler with the configured settings
+            MqttClientSslData sslData = clientData.getSslData().orElseThrow(() -> new IllegalStateException("SSL used, but no sslData present"));
+            final SslHandler sslHandler = createSslHandler(channel, sslData);
+            //add the handler as first handler to the pipeline
+            channel.pipeline().addFirst(sslHandler);
```

Comment:
Is SSL without websockets working? I thought we had to add a listener to the handshakeFuture and call initMqttHandlers there

Reply:
yes, should be working. I had originally coded with a listener, but it didn't work because of events that were then already consumed but on which the MQTT handlers wanted to react. The original code was something like this:

```
    @Override
    protected void initChannel(final SocketChannel channel) {
        channelComponent = ChannelComponent.create(channel, clientData);
        if (clientData.usesSSL()) {
            initSsl(channel);
        } else {
            init(channel);
        }
    }

    private void initSsl(final SocketChannel channel) {
        final Optional<MqttClientSslData> sslDataOptional = clientData.getSslData();
        if (!sslDataOptional.isPresent()) {
            //throw new IllegalStateException("SSL used, but no sslData present");
        }

        try {
            //create a new SslHandler with the configured settings
            final SslHandler sslHandler = createSslHandler(channel, sslDataOptional.get());
            sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<? super Channel>>() {
                @Override
                public void operationComplete(Future<? super Channel> future) throws Exception {
                    if (future.isDone() && future.isSuccess()) {
                        init((Channel) future.get());
                    }
                }
            });

            //add the handler as first handler to the pipeline
            channel.pipeline().addFirst(sslHandler);

        } catch (final SSLException e) {
            channel.pipeline().fireExceptionCaught(e);
        }
    }
```

init() then called initMqttHandlers

Reply:
If you look at the SmokeTest, you can see that all the various configurations are working.